File: astar/astar.cpp
#include <stdio.h>
#include <stdlib.h>
#include <iostream>
#include <vector>
#include <queue>
#include <map>
#include <sstream>
using namespace std;


int width = 10, height = 10;
int cfg[10][10] = {
  {1, 0, 0, 0, 0, 0, 0, 0, 0, 0},
  {0, 1, 0, 0, 0, 0, 0, 0, 0, 0},
  {0, 0, 1, 0, 0, 0, 0, 0, 0, 0},
  {0, 0, 0, 1, 0, 0, 0, 0, 0, 0},
  {0, 0, 0, 0, 1, 0, 0, 0, 0, 0},
  {0, 0, 0, 0, 0, 1, 0, 0, 0, 0},
  {0, 0, 0, 0, 0, 0, 1, 0, 0, 0},
  {0, 0, 0, 0, 0, 0, 0, 1, 0, 0},
  {0, 0, 0, 0, 0, 0, 0, 0, 1, 0},
  {0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
};

typedef struct Position{
  Position(){}
  Position(int x, int y)
  {
    this->x = x;
    this->y = y;
  }

  int x;
  int y;

  inline bool operator==(Position& b)
  {
    return x == b.x && y == b.y;
  }
}Position;

typedef struct Node{
  Node(){}
  Node(const Position& pos)
  {
    this->pos = pos;
    this->id = pos.y * width + pos.x;
    H = 0;
    G = 0;
    F = 0;
    closed = false;
  }

  void setH(const Position& pos){
    this->H = abs(this->pos.x - pos.x) + abs(this->pos.y - pos.y);
  }

  void setG(int v){
    this->G = v;
    this->F = G + H;
  }

  int F;
  int G;  //当前的消耗
  int H;  //距离终点的预估值
  int id;
  bool closed;
  Position pos;
  Node* parent;

  // inline bool operator<(const Node& b)
  // {
  //   return F < b.F;
  // }
}Node, *pNode;


Node* crete_new_node(const Position& pos, map<int, pNode>& list, const Position& end){
  auto node = new Node(pos);
  node->setH(end);
  list[node->id] = node;
  return node;
}

int main(int argc, char* argv[])
{
  auto comp = [=](pNode a, pNode b) { return a->F > b->F; };
  priority_queue<pNode, vector<pNode>, decltype(comp) > open_list(comp);//开放列表
  map<int, pNode> open_list_map;
  map<int, pNode> close_list;                                           //关闭列表
  map<int, pNode> node_list;                                            //全部节点列表

  pNode find = nullptr;
  Position start(0,2);
  Position end(9, 2);
  pNode s = crete_new_node(start, node_list, end);
  open_list.push(s);//起点加入开放列表
  open_list_map[s->id] = s;

  while(!open_list.empty())
  {
    auto current = open_list.top();
    close_list[current->id] = current;//加入关闭列表
    open_list.pop();
    open_list_map.erase(current->id);


    auto f = current->F;
    for (auto v:open_list_map) 
    {
      if (f > v.second->F)
      {
        int test = 0;
      }
    }


    if (current->pos == end){
      find = current;
      break;
    }

    int dir[][2] = { {0,1}, {0,-1}, {1,0}, {-1,0}};
    for (int i=0; i < 4; i++)
    {
      auto v = dir[i];
      int x = current->pos.x + v[0];
      int y = current->pos.y + v[1];
      int id = y * width + x;
      if (x < 0 || x >= width || y < 0 || y >= height || cfg[x][y] != 0) continue;
      if (close_list.find(id) != close_list.end()) continue;

      pNode next_node = nullptr;
      if (open_list_map.find(id) != open_list_map.end())
      {
        next_node = open_list_map.at(id);
        if (current->G + 10 < next_node->G)
        {
          next_node->setG(current->G + 10);
          next_node->parent = current;
        }
      }
      else
      {
        next_node = crete_new_node(Position(x, y), node_list, end);
        next_node->parent = current;
        next_node->setG(current->G + 10);
        open_list.push(next_node);
        open_list_map[next_node->id] = next_node;
      }
    }
  }

  vector<pNode> path;
  while(find != nullptr)
  {
    path.push_back(find);
    find = find->parent;
  }

  for (auto node : path)
  {
    cfg[node->pos.x][node->pos.y] = 2;
  }
  for (int y = 0; y < height; y++)
  {
    stringstream fmt;
    fmt << "lien:" << y << " ";
    for (int x = 0; x < width; x++)
    {
      fmt << cfg[x][y] << " ";
    }

    cout << fmt.str() << endl;
  }

  //清理
  for (auto v : node_list)
  {
    delete v.second;
  }

  return 0;
}

File: astar/astar.lua
--0: 可通过 1:障碍物
local cfg = {
  {1, 0, 0, 0, 0, 0, 0, 0, 0, 0}, --1
  {0, 1, 0, 0, 0, 0, 0, 0, 0, 0}, --2
  {0, 0, 1, 0, 0, 0, 0, 0, 0, 0}, --3
  {0, 0, 0, 1, 0, 0, 0, 0, 0, 0}, --4
  {0, 0, 0, 0, 1, 0, 0, 0, 0, 0}, --5
  {0, 0, 0, 0, 0, 1, 0, 0, 0, 0}, --6
  {0, 0, 0, 0, 0, 0, 0, 0, 0, 0}, --7
  {0, 0, 0, 0, 0, 0, 0, 0, 0, 0}, --8
  {0, 0, 0, 0, 0, 0, 0, 0, 0, 0}, --9
  {0, 0, 0, 0, 0, 0, 0, 0, 0, 0}, --10
}
local width, height = 10,10

local function gid(pos)
  return pos.y * width + pos.x
end
local function position(x, y)
  return {x=x, y=y}
end

local function new_node(pos)
  local t = {}
  t.pos = pos
  t.parent = nil
  t.G = 0 --当前的消耗
  t.H = 0 --预估的消耗
  t.F = 0
  t.id = gid(pos)
  return t
end 

local function sort(a, b)
  return a.F < b.F
end

local function find_from_open_list(list, id)
  for _,v in pairs(list) do
    if v.id == id then
      return v
    end
  end
end

local function position_equal(p1, p2)
  return p1.x == p2.x and p1.y == p2.y
end
local function position_valid(x, y)
  return x<=width and x>=1 and y<= height and y>= 1 and cfg[x][y] == 0
end

local start = new_node(position(1, 5))
local goal  = new_node(position(10, 5))
local close_list = {}
local open_list = {[1] = start}
local find = nil


while (#open_list ~= 0) do
  table.sort(open_list, sort)
  local current = open_list[1]--取F最小的作为当前节点
  table.remove(open_list, 1)  --从open_list中移除
  close_list[current.id] = current--添加到关闭列表
  if current.id == goal.id then
    find = current
    break
  end

  --只取周围4个方向
  for _,dir in ipairs({ {1,0}, {-1,0}, {0,1}, {0,-1} }) do
    local x,y = current.pos.x + dir[1], current.pos.y + dir[2]
    local id = gid(position(x, y))
    if position_valid(x, y) and close_list[id] == nil then--有效节点而且不在关闭列表中
      local next_node = find_from_open_list(open_list, id)
      if next_node ~= nil then  --已经在开放列表中
        --检查经过这个节点到达 是否F值会更低
        if current.G + 10 < next_node.G then
          next_node.parent = current
          next_node.G = current.G + 10
          next_node.F = next_node.G + next_node.H
        end
      else                      --加入开放列表
        next_node = new_node(position(x, y))
        next_node.parent = current
        next_node.G = current.G + 10
        next_node.F = next_node.G + next_node.H
        table.insert(open_list, next_node)
      end
    end
  end
end

local path = {}
local tmp = {}
local node = find
while node ~= nil do
  table.insert(tmp, node)
  node = node.parent
end
for i=#tmp,1,-1 do
  table.insert(path, tmp[i])
end

local path_map = {}
for _,v in ipairs(path) do
  local id = gid(v.pos)
  path_map[id] = 6
end
local function get_value(x, y)
  local id = gid(position(x, y))
  return path_map[id] or cfg[x][y]
end

for y,line in ipairs(cfg) do
  local value = ''
  for x,_ in ipairs(line) do
    value = value .. get_value(x, y) .. '  '
  end
  print('line'..y..': ' .. value)
end